### PR TITLE
Akka http client

### PR DESCRIPTION
**Pull request by anilgursel**
Thanks for your pull request.  Please review the following guidelines.

- [x] Title includes issue id.
- [x] Description of the change added.
- [x] Commits are squashed.
- [x] Tests added.
- [ ] Documentation added/updated.
- [x] Also please review [CONTRIBUTING.md](https://github.com/paypal/squbs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: squbs-httpclient/src/main/scala/org/squbs/endpoint/Endpoint.scala [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+  def name: String
+  def resolve(svcName: String, env: Environment = Default): Option[Endpoint]
+}
+
+class EndpointResolverRegistryExtension(system: ExtendedActorSystem) extends Extension with LazyLogging {
+  val endpointResolvers = ListBuffer[EndpointResolver]()
+
+  def register(resolver: EndpointResolver) = {
+    endpointResolvers.find(_.name == resolver.name) match {
+      case None => endpointResolvers.prepend(resolver)
```
Comment (akara) [inline squbs-httpclient/src/main/scala/org/squbs/endpoint/Endpoint.scala]: I'd use a `List` and just say `resolver :: endpointResolvers`. This is based on the search pattern.

File: squbs-httpclient/src/main/scala/org/squbs/endpoint/Endpoint.scala [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    endpointResolvers.find(_.resolve(svcName, env) != None)
+  }
+
+  def resolve(svcName: String, env: Environment = Default): Option[Endpoint] = {
+    val resolvedEndpoint = endpointResolvers.foldLeft[Option[Endpoint]](None) {
+      (endpoint: Option[Endpoint], resolver: EndpointResolver) =>
+        endpoint match {
+          case Some(_) => endpoint
+          case None => resolver.resolve(svcName, env)
+        }
```
Comment (akara) [inline squbs-httpclient/src/main/scala/org/squbs/endpoint/Endpoint.scala]: Looking at this logic, if we have a long resolver chain but most of the resolutions end up on the first resolver (say topo), this seems to keep the match going until all resolvers are exhausted. A `find` would perhaps be better?

File: squbs-httpclient/src/main/scala/org/squbs/env/Environment.scala [hunk 3]
```diff
@@ -89,15 +89,15 @@ class EnvironmentRegistryExtension(system: ExtendedActorSystem) extends Extensio
     }
   }
 
-  def resolve(svcName: String): Environment = {
+  def resolve: Environment = {
     val resolvedEnv = environmentResolvers.foldLeft[Environment](Default){
       (env: Environment, resolver: EnvironmentResolver) =>
         env match {
-          case Default => resolver.resolve(svcName)
+          case Default => resolver.resolve
```
Comment (akara) [inline squbs-httpclient/src/main/scala/org/squbs/env/Environment.scala]: Ditto here.

File: squbs-httpclient/src/main/scala/org/squbs/httpclient/ClientFlow.scala [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+  def pipelineAdapter(clientConnectionFlow: ClientConnectionFlow[RequestContext]):
+  Flow[RequestContext, RequestContext, HostConnectionPool] = {
+    val fromRc = Flow[RequestContext].map { rc => (rc.request, rc) }
+    val toRc = Flow[(Try[HttpResponse], RequestContext)].map {
+      case (responseTry, rc) => rc.copy(response = Option(responseTry))
+    }
+
+    fromRc.viaMat(clientConnectionFlow)(Keep.right).via(toRc)
+  }
+}
```
Comment (akara) [inline squbs-httpclient/src/main/scala/org/squbs/httpclient/ClientFlow.scala]: I think I saw this code before. LGTM.

File: squbs-streamingpipeline/src/main/scala/org/squbs/pipeline/streaming/PipelineExtension.scala [hunk 5]
```diff
@@ -23,30 +23,36 @@ import com.typesafe.config.ConfigObject
 
 import scala.annotation.tailrec
 
+case class Context(name: String)
+
 trait PipelineFlowFactory {
 
-  def create(implicit system: ActorSystem): BidiFlow[RequestContext, RequestContext, RequestContext, RequestContext, NotUsed]
+  def create(context: Context)(implicit system: ActorSystem):
```
Comment (anilgursel) [inline squbs-streamingpipeline/src/main/scala/org/squbs/pipeline/streaming/PipelineExtension.scala]: Instead of `Context`, I can also do the following:

```scala
object MyObject {
  val w1 = Witness("title")
  val w2 = Witness("id")
  val w3 = Witness("foo3")

  type HasFoo1[L <: HList] = Selector[L, w1.T] { type Out = String }
  type HasFoo2[L <: HList] = Selector[L, w2.T]
  type HasFoo3[L <: HList] = Selector[L, w3.T]

  @implicitNotFound("${L} should have foo1, foo2 and foo3")
  case class HasMyFields[L <: HList](implicit s1: HasFoo1[L], s2: HasFoo2[L], s3: HasFoo3[L])

  object HasMyFields {
    implicit def make[L <: HList : HasFoo1 : HasFoo2 : HasFoo3] = HasMyFields[L]
  }
}

// Then, the function can be defined as 

def create[L <: HList : HasMyFields](xs: L)(implicit system: ActorSystem): BidiFlow[RequestContext, RequestContext, RequestContext, RequestContext, NotUsed]
```

An implementation woud look like:

```scala
def create[L <: HList : HasMyFields](xs: L)(implicit system: ActorSystem): BidiFlow[RequestContext, RequestContext, RequestContext, RequestContext, NotUsed] = {
  val selectors = implicitly[HasMyFields[L]]
  import selectors._
  xs("title"), xs("id")
  // And create BidiFlow..
}
```

Two downsides of this approach:

   * Too Scala oriented.  Need to think about Java.
   * Requires the following inside the implementation function..  I can actually try to get this cleaner:
      ```scala
      val selectors = implicitly[HasMyFields[L]]
      import selectors._
      ```

File: squbs-httpclient/src/main/scala/org/squbs/endpoint/Endpoint.scala [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+  def create(s: String): Endpoint = Endpoint(new URI(s))
+}
+
+trait EndpointResolver {
+  def name: String
+  def resolve(svcName: String, env: Environment = Default): Option[Endpoint]
+}
+
+class EndpointResolverRegistryExtension(system: ExtendedActorSystem) extends Extension with LazyLogging {
+  val endpointResolvers = ListBuffer[EndpointResolver]()
```
Comment (anilgursel) [inline squbs-httpclient/src/main/scala/org/squbs/endpoint/Endpoint.scala]: 👍

File: squbs-httpclient/src/main/scala/org/squbs/endpoint/Endpoint.scala (hunk 1 above)
Comment (anilgursel) [inline squbs-httpclient/src/main/scala/org/squbs/endpoint/Endpoint.scala]: 👍

File: squbs-httpclient/src/main/scala/org/squbs/endpoint/Endpoint.scala (hunk 2 above)
Comment (anilgursel) [inline squbs-httpclient/src/main/scala/org/squbs/endpoint/Endpoint.scala]: This is a copy paste.. and actually the readability of the `fold` is not as clean, IMO..

On the other side, not sure how `find` would work.  It would require me to call `resolve` twice:

```scala
val resolvedEndpoint = endpointResolvers.find(_.resolve(svcName, env).nonEmpty).flatMap(_.resolve(svcName, env))
```

How about the following?

```scala
val resolvedEndpoint = endpointResolvers.view map {_.resolve(svcName, env)} collectFirst { case Some(x) => x }
```

File: squbs-httpclient/src/main/scala/org/squbs/httpclient/HttpClientCircuitBreaker.scala [hunk 7]
```diff
@@ -39,31 +38,6 @@ object ServiceCallStatus extends Enumeration {
...
-      }
-      case Failure(e) => client.cbMetrics.add(ServiceCallStatus.Exception, System.nanoTime)
-    }
-    fallbackHttpResponse map { fallback =>
-      runCircuitBreaker recover {
-        case e: CircuitBreakerOpenException => fallback
-      }
-    } getOrElse runCircuitBreaker
-  }
-}
```
Comment (akara) [inline squbs-httpclient/src/main/scala/org/squbs/httpclient/HttpClientCircuitBreaker.scala]: Pending having the circuit breaker design, right?

File: squbs-httpclient/src/main/scala/org/squbs/httpclient/HttpClientJMX.scala [hunk 8]
```diff
@@ -47,7 +47,7 @@ object HttpClientJMX {
 
   def registryHCBean(implicit system: ActorSystem) = {
     if (!JMX.isRegistered(httpClientBean)){
-      JMX.register(HttpClientBean(system), JMX.prefix(system) + httpClientBean)
+//      JMX.register(HttpClientBean(system), JMX.prefix(system) + httpClientBean)
```
Comment (akara) [inline squbs-httpclient/src/main/scala/org/squbs/httpclient/HttpClientJMX.scala]: Temporary?

File: squbs-httpclient/src/main/scala/org/squbs/metrics/MetricsExtension.scala [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+  /**
+    * Java API: retrieve the Count extension for the given system.
+    */
+  override def get(system: ActorSystem): MetricsExtensionImpl = super.get(system)
+}
+
+class MetricsExtensionImpl(system: ActorSystem) extends Extension {
+  val Domain = s"org.squbs.metrics.${system.name}"
+  val metrics = new MetricRegistry()
```
Comment (akara) [inline squbs-httpclient/src/main/scala/org/squbs/metrics/MetricsExtension.scala]: `MetricRegistry` or `MetricsRegistry`?

File: squbs-httpclient/src/test/scala/org/squbs/endpoint/EndpointSpec.scala [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+    a[RuntimeException] should be thrownBy {
+      EndpointResolverRegistry(system).register(new DummyLocalhostResolver)
+      EndpointResolverRegistry(system).route("abcService", QA)
+    }
+  }
+
+  it should "resolve the endpoint when Environment is provided" in {
+    EndpointResolverRegistry(system).register(new DummyLocalhostResolver)
+    EndpointResolverRegistry(system).route("abcService", DEV) should not be None
+    EndpointResolverRegistry(system).route("abcService", DEV).get.name should be ("DummyLocalhostResolver")
```
Comment (akara) [inline squbs-httpclient/src/test/scala/org/squbs/endpoint/EndpointSpec.scala]: Does `route` return an `Option`? If so, please use the ScalaTest option values `.value` instead of `.get`. The error behavior is different between the two.

File: squbs-httpclient/src/test/scala/org/squbs/endpoint/EndpointSpec.scala [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+    EndpointResolverRegistry(system).register(new EndpointResolver {
+      override def resolve(svcName: String, env: Environment = Default): Option[Endpoint] =
+        Some(Endpoint("http://localhost:9090"))
+
+      override def name: String = "override"
+    })
+    EndpointResolverRegistry(system).endpointResolvers should have size 2
+    EndpointResolverRegistry(system).endpointResolvers.head should not be a [DummyLocalhostResolver]
+    EndpointResolverRegistry(system).endpointResolvers.head.name should be ("override")
+    EndpointResolverRegistry(system).route("abcService") should not be None
```
Comment (akara) [inline squbs-httpclient/src/test/scala/org/squbs/endpoint/EndpointSpec.scala]: I think we should use `should be 'defined` in this case.

File: squbs-httpclient/src/test/scala/org/squbs/endpoint/EndpointSpec.scala [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+    EndpointResolverRegistry(system).route("abcService").get.name should be ("override")
```
Comment (akara) [inline squbs-httpclient/src/test/scala/org/squbs/endpoint/EndpointSpec.scala]: Ditto with `.value`. Please check the ScalaTest matcher doc.

File: squbs-httpclient/src/test/scala/org/squbs/endpoint/EndpointSpec.scala [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+          case "unique" => Some(Endpoint("http://www.ebay.com"))
+          case _ => None
+        }
+      }
+
+      override def name: String = "unique"
+    })
+    EndpointResolverRegistry(system).endpointResolvers should have size 2
+    EndpointResolverRegistry(system).route("abcService") should not be None
+    EndpointResolverRegistry(system).route("abcService").get.name should be ("DummyLocalhostResolver")
```
Comment (akara) [inline squbs-httpclient/src/test/scala/org/squbs/endpoint/EndpointSpec.scala]: Ditto all the places with `Option.get`. Codacy will red-mark this code.

File: squbs-httpclient/src/test/scala/org/squbs/httpclient/ClientFlowPipelineSpec.scala [hunk 14]
```diff
@@ -0,0 +1,214 @@
...
+      RawHeader("keyPreInbound", "valPreInbound"),
+      RawHeader("keyPostInbound", "valPostInbound")).sortBy(_.name).mkString(",")
+
+    assertPipeline("clientWithCustomPipelineWithDefaults", expectedResponseHeaders, expectedEntity)
+  }
+
+  it should "build the flow only with defaults" in {
+    val expectedResponseHeaders = Seq(
+      RawHeader("keyPreOutbound", "valPreOutbound"),
+      RawHeader("keyPostOutbound", "valPostOutbound")).sortBy(_.name)
```
Comment (akara) [inline squbs-httpclient/src/test/scala/org/squbs/httpclient/ClientFlowPipelineSpec.scala]: Do we sort in order to compare? If so, we can use the DSL `contain theSameElementsAs` instead.

File: squbs-httpclient/src/test/scala/org/squbs/httpclient/ClientFlowPipelineSpec.scala [hunk 15]
```diff
@@ -0,0 +1,214 @@
...
+
+    val expectedEntity =  Seq(
+      RawHeader("keyPreInbound", "valPreInbound"),
+      RawHeader("keyPostInbound", "valPostInbound")).sortBy(_.name).mkString(",")
```
Comment (akara) [inline squbs-httpclient/src/test/scala/org/squbs/httpclient/ClientFlowPipelineSpec.scala]: Ditto with the sort.

File: squbs-httpclient/src/test/scala/org/squbs/httpclient/ClientFlowSpec.scala [hunk 16]
```diff
@@ -0,0 +1,180 @@
...
+    val cps = ClientFlow.connectionPoolSettings("sampleclient", config, None)
+    cps.maxConnections shouldBe 987
+    cps.maxRetries shouldBe 123
+    cps.idleTimeout.toSeconds should equal (defaultConfig.getDuration("akka.http.host-connection-pool.idle-timeout").getSeconds)
+
+    val cps2 = ClientFlow.connectionPoolSettings("sampleclient2", config, None)
+    cps2.maxConnections shouldBe 666
+    cps2.maxRetries shouldBe defaultConfig.getInt("akka.http.host-connection-pool.max-retries")
+    cps2.idleTimeout.toSeconds should equal (defaultConfig.getDuration("akka.http.host-connection-pool.idle-timeout").getSeconds)
+
```
Comment (akara) [inline squbs-httpclient/src/test/scala/org/squbs/httpclient/ClientFlowSpec.scala]: We should discuss the config structure a bit. I'm fine with the code but would like to get more clarity there.

File: squbs-httpclient/src/test/scala/org/squbs/metrics/MetricsFlowSpec.scala [hunk 17]
```diff
@@ -0,0 +1,223 @@
...
+      assertJmxValue("sampleClient5-4XX-count", "Count", 1)
+    }
+  }
+
+  it should "collect metrics for exceptions" in {
+    val f = for {
+      hello <- callServiceTwice("sampleClient6", "/hello")
+      connectionException <- callService("sampleClient6", "/connectionException")
+      connectionException2 <- callService("sampleClient6", "/connectionException")
+      timeoutException <- callServiceTwice("sampleClient6", "/timeoutException")
```
Comment (akara) [inline squbs-httpclient/src/test/scala/org/squbs/metrics/MetricsFlowSpec.scala]: Is there anything we can do at the pipeline level if some `FlowDefinition` collapses the stream? Not sure of that.

File: squbs-streamingpipeline/src/main/scala/org/squbs/pipeline/streaming/PipelineExtension.scala [hunk 18]
```diff
@@ -71,18 +77,18 @@ object PipelineExtension extends ExtensionId[PipelineExtensionImpl] with Extensi
...
 
-    var flowMap = Map.empty[String, PipelineFlow]
+    var flowMap = Map.empty[String, PipelineFlowFactory]
     flows foreach { case (name, config) =>
       val factoryClassName = config.getString("factory")
 
       val flowFactory = Class.forName(factoryClassName).newInstance().asInstanceOf[PipelineFlowFactory]
 
-      flowMap = flowMap + (name -> flowFactory.create(system))
+      flowMap = flowMap + (name -> flowFactory)
```
Comment (akara) [inline squbs-streamingpipeline/src/main/scala/org/squbs/pipeline/streaming/PipelineExtension.scala]: Ah, you keep the `FlowFactory` instance instead of the `PipelineFlow` instance.

File: squbs-streamingpipeline/src/test/scala/org/squbs/pipeline/streaming/RequestContextSpec.scala [hunk 19]
```diff
@@ -65,13 +67,13 @@ class RequestContextSpec extends TestKit(ActorSystem("RequestContextSpecSys")) w
     rc3.response should be(None)
     rc3.request.headers should contain(RawHeader("key1", "val1"))
 
-    val rc4 = rc3.copy(response = Some(HttpResponse()))
-    rc4.response.get.headers should have size 0
+    val rc4 = rc3.copy(response = Some(Try(HttpResponse())))
+    rc4.response.get.get.headers should have size 0
```
Comment (akara) [inline squbs-streamingpipeline/src/test/scala/org/squbs/pipeline/streaming/RequestContextSpec.scala]: Don't like all these `.get`s. Please do `.value` if possible. Again, the failure mode is different.

File: squbs-streamingpipeline/src/test/scala/org/squbs/pipeline/streaming/RequestContextSpec.scala [hunk 20]
```diff
@@ -65,13 +67,13 @@ class RequestContextSpec extends TestKit(ActorSystem("RequestContextSpecSys")) w
...
     val rc5 = rc4.addResponseHeader(RawHeader("key4", "val4"))
     val rc6 = rc5.addResponseHeaders(RawHeader("key5", "val5"), RawHeader("key6", "val6"))
     val rc7 = rc6.addResponseHeader(RawHeader("key7", "val7"))
     rc7.request.headers should have size 3
-    rc7.response.get.headers should have size 4
-    rc7.response.get.headers should contain(RawHeader("key4", "val4"))
+    rc7.response.get.get.headers should have size 4
+    rc7.response.get.get.headers should contain(RawHeader("key4", "val4"))
```
Comment (akara) [inline squbs-streamingpipeline/src/test/scala/org/squbs/pipeline/streaming/RequestContextSpec.scala]: Ditto.

File: squbs-unicomplex/src/main/scala/org/squbs/unicomplex/streaming/FlowHandler.scala [hunk 21]
```diff
@@ -123,14 +126,14 @@ class FlowHandler(routes: Seq[(Path, ServiceHandlerWrapper, PipelineSetting)], l
...
-        pipelineExtension.getFlow(pipelineSettings(i)) match {
+        pipelineExtension.getFlow(pipelineSettings(i), Context(name = paths(i).toString())) match {
           case Some(pipeline) => partition.out(i) ~> pipeline.join(serviceFlow) ~> routesMerge
           case None => partition.out(i) ~> serviceFlow ~> routesMerge
         }
       }
 
       val notFound = b.add(Flow[RequestContext].map {
-        rc => rc.copy(response = Option(notFoundHttpResponse))
+        rc => rc.copy(response = Option(Try(NotFound)))
```
Comment (akara) [inline squbs-unicomplex/src/main/scala/org/squbs/unicomplex/streaming/FlowHandler.scala]: No problem with your code. Just want to let you know there will be some merge work on this file. I hope it's not too bad.

File: squbs-streamingpipeline/src/main/scala/org/squbs/pipeline/streaming/PipelineExtension.scala (hunk 5 above)
Comment (anilgursel) [inline squbs-streamingpipeline/src/main/scala/org/squbs/pipeline/streaming/PipelineExtension.scala]: Create https://github.com/paypal/squbs/issues/307 to come back to it later.

File: squbs-httpclient/src/main/scala/org/squbs/httpclient/HttpClientCircuitBreaker.scala (hunk 7 above)
Comment (anilgursel) [inline squbs-httpclient/src/main/scala/org/squbs/httpclient/HttpClientCircuitBreaker.scala]: Yes.  Still do not have a good design and option from stream :(

File: squbs-httpclient/src/main/scala/org/squbs/httpclient/HttpClientJMX.scala (hunk 8 above)
Comment (anilgursel) [inline squbs-httpclient/src/main/scala/org/squbs/httpclient/HttpClientJMX.scala]: Yes this is temporary.  Actually, this will be replaced with new set of JMX registrations.

File: squbs-httpclient/src/main/scala/org/squbs/metrics/MetricsExtension.scala (hunk 9 above)
Comment (anilgursel) [inline squbs-httpclient/src/main/scala/org/squbs/metrics/MetricsExtension.scala]: `com.codahale.metrics.MetricRegistry`.  The question is, should I rename squbs classes as `Metrics*` or `Metric*`.

File: squbs-httpclient/src/test/scala/org/squbs/endpoint/EndpointSpec.scala (hunk 10 above)
Comment (anilgursel) [inline squbs-httpclient/src/test/scala/org/squbs/endpoint/EndpointSpec.scala]: Thanks!

File: squbs-httpclient/src/test/scala/org/squbs/endpoint/EndpointSpec.scala (hunk 11 above)
Comment (anilgursel) [inline squbs-httpclient/src/test/scala/org/squbs/endpoint/EndpointSpec.scala]: 👍

File: squbs-httpclient/src/test/scala/org/squbs/httpclient/ClientFlowPipelineSpec.scala (hunk 14 above)
Comment (anilgursel) [inline squbs-httpclient/src/test/scala/org/squbs/httpclient/ClientFlowPipelineSpec.scala]: Actually, the assertion already uses `contain theSameElementsAs`..  No need sorting here..

File: squbs-httpclient/src/test/scala/org/squbs/httpclient/ClientFlowSpec.scala (hunk 16 above)
Comment (anilgursel) [inline squbs-httpclient/src/test/scala/org/squbs/httpclient/ClientFlowSpec.scala]: Sure.  Here is how to override a config specifically for a client:

```
sampleclient {
 type = squbs.httpclient

 akka.http {
   host-connection-pool {
     max-connections = 987
     max-retries = 123

     client = {
       connecting-timeout = 123 ms
     }
   }
 }
}
```

File: squbs-httpclient/src/test/scala/org/squbs/metrics/MetricsFlowSpec.scala (hunk 17 above)
Comment (anilgursel) [inline squbs-httpclient/src/test/scala/org/squbs/metrics/MetricsFlowSpec.scala]: Not clear about the comment.  Would you please elaborate?

File: squbs-streamingpipeline/src/test/scala/org/squbs/pipeline/streaming/RequestContextSpec.scala (hunk 19 above)
Comment (anilgursel) [inline squbs-streamingpipeline/src/test/scala/org/squbs/pipeline/streaming/RequestContextSpec.scala]: Will do!

File: squbs-unicomplex/src/main/scala/org/squbs/unicomplex/streaming/FlowHandler.scala (hunk 21 above)
Comment (anilgursel) [inline squbs-unicomplex/src/main/scala/org/squbs/unicomplex/streaming/FlowHandler.scala]: Yeah, I expect some merge fun :)

File: squbs-httpclient/src/main/scala/org/squbs/metrics/MetricsExtension.scala (hunk 9 above)
Comment (akara) [inline squbs-httpclient/src/main/scala/org/squbs/metrics/MetricsExtension.scala]: I'd go by the merit. If the class is a single metric, use the singular form. If it is multiple, we should use the plural form. Keeps it clean. If we want to change it, now is the time. But we can do that in a separate PR. These mega-PRs are hard to review. Separate ones with concise purpose are definitely better. These ones are just sometimes unavoidable.

File: squbs-httpclient/src/test/scala/org/squbs/httpclient/ClientFlowSpec.scala (hunk 16 above)
Comment (akara) [inline squbs-httpclient/src/test/scala/org/squbs/httpclient/ClientFlowSpec.scala]: Override? What is the fallback path if not overridden?

File: squbs-httpclient/src/test/scala/org/squbs/metrics/MetricsFlowSpec.scala (hunk 17 above)
Comment (akara) [inline squbs-httpclient/src/test/scala/org/squbs/metrics/MetricsFlowSpec.scala]: Just roughly:

Request ~> Pipeline ~> AppFlow ~> Pipeline ~> Response

If AppFlow causes the stream to totally collapse due to a thrown exception, I'd like to catch it in the pipeline and perhaps send a more meaningful response than just a blank 500 that is the case today.

**Comment by akara** [comment]
LGTM for this PR. The comments can be addressed separately. They are a bit outside the merit of this PR, which is a big one in a series of PRs to come.

File: squbs-httpclient/src/test/scala/org/squbs/httpclient/ClientFlowSpec.scala (hunk 16 above)
Comment (anilgursel) [inline squbs-httpclient/src/test/scala/org/squbs/httpclient/ClientFlowSpec.scala]: For any setting `s`, e.g., `akka.http.host-connection-pool.max-connections`, we first check `httpclientname.s` and then `s`.  For example, if `sampleclient.akka.http.host-connection-pool.max-connections` does not exist, `akka.http.host-connection-pool.max-connections` is the value..  They way I implement it is basically getting a `Config` object `c` for "sampleclient", then use `c.withFallback(system.settings.config)`..  Then, pass that to Akka Http APIs to setup pool settings.

File: squbs-httpclient/src/test/scala/org/squbs/metrics/MetricsFlowSpec.scala (hunk 17 above)
Comment (anilgursel) [inline squbs-httpclient/src/test/scala/org/squbs/metrics/MetricsFlowSpec.scala]: I see, you are talking about the server side.

I do not think we can catch an exception in the pipeline.  We can register a default `decider` for the flow though that can create a `Try` and send it as a message.  Then, we can try to integrate exception mappers..  Basically, if failed, instead of just doing 500, we can try to first do a mapping.

I think it deserves an issue https://github.com/paypal/squbs/issues/312.
